mlflow/pydantic_ai/autolog.py

```diff
@@ -11,19 +11,88 @@
 
 _logger = logging.getLogger(__name__)
 
+# Allowlists for safe attributes to extract from pydantic_ai objects.
+# Using allowlists instead of denylists to avoid capturing client/provider
+# references that can interfere with async cleanup (e.g., httpx client lifecycle).
+_AGENT_SAFE_ATTRIBUTES = frozenset(
+    {
+        "name",
+        "system_prompt",
+        "retries",
+        "result_type",
+        "output_type",
+        "deps_type",
+        "end_strategy",
+        "defer_model_check",
+        "instrument",
+    }
+)
+
+_MODEL_SAFE_ATTRIBUTES = frozenset(
+    {
+        "model_name",
+        "name",
+        "system",
+    }
+)
+
+_TOOL_SAFE_ATTRIBUTES = frozenset(
+    {
+        "name",
+        "description",
+        "max_retries",
+        "prepare",
+    }
+)
+
+_MCP_SERVER_SAFE_ATTRIBUTES = frozenset(
+    {
+        "name",
+        "url",
+    }
+)
+
+_SAFE_PRIMITIVE_TYPES = (str, int, float, bool)
+
+
+def _is_safe_for_serialization(value: Any) -> bool:
+    if value is None:
+        return False
+    if isinstance(value, _SAFE_PRIMITIVE_TYPES):
+        return True
+    if isinstance(value, dict):
+        return all(_is_safe_for_serialization(v) for v in value.values())
+    if isinstance(value, (list, tuple)):
+        return all(_is_safe_for_serialization(v) for v in value)
+    if hasattr(value, "__dataclass_fields__"):
+        return True
```

Comment:
Is this always true?

Reply:
There's a cleaner way to do this... instead of using dunder commands I'll use the standard method just as a guard to protect unserializable subclass components from the library